Mathlib/Order/Partition/Basic.lean

```diff
@@ -0,0 +1,118 @@
+  nontrivial elements whose supremum is `s`.
+
+-/
+variable {α : Type*} {s x y z : α}
+
+open Set
+
+/-- A `Partition` of an element `s` of a `CompleteLattice` is a collection of
+independent nontrivial elements whose supremum is `s`.  -/
+structure Partition [CompleteLattice α] (s : α) where
+  /-- The collection of parts-/
+  parts : Set α
+  /-- The parts are `sSupIndep`-/
+  indep : sSupIndep parts
+  /-- The bottom element is not a part-/
+  bot_not_mem : ⊥ ∉ parts
+  /-- The supremum of all parts is `s`-/
+  sSup_parts_eq : sSup parts = s
+
+namespace Partition
+
+section Basic
+
+variable [CompleteLattice α] {P : Partition s}
+
+instance {α : Type*} [CompleteLattice α] {s : α} : SetLike (Partition s) α where
+  coe := Partition.parts
+  coe_injective' p p' h := by cases p; cases p'; simpa using h
+
+@[simp] lemma coe_parts : (P.parts : Set α) = P := rfl
+
+@[ext] lemma ext {P Q : Partition s} (hP : ∀ x, x ∈ P ↔ x ∈ Q) : P = Q :=
+  SetLike.ext hP
+
+lemma disjoint (hx : x ∈ P) (hy : y ∈ P) (hxy : x ≠ y) :
+    Disjoint x y :=
+  P.indep.pairwiseDisjoint hx hy hxy
+
+lemma pairwiseDisjoint : Set.PairwiseDisjoint (P : Set α) id :=
+  fun _ hx _ hy hxy ↦ P.disjoint hx hy hxy
+
+@[simp]
+lemma sSup_eq (P : Partition s) : sSup P = s :=
+  P.sSup_parts_eq
+
+@[simp]
+lemma iSup_eq (P : Partition s) : ⨆ x ∈ P, x = s := by
+  simp_rw [← P.sSup_parts_eq, sSup_eq_iSup]
+  rfl
+
+lemma le_of_mem (P : Partition s) (hx : x ∈ P) : x ≤ s :=
+  (le_sSup hx).trans_eq P.sSup_eq
+
+lemma parts_nonempty (P : Partition s) (hs : s ≠ ⊥) : (P : Set α).Nonempty :=
+  nonempty_iff_ne_empty.2 fun hP ↦ by simp [← P.sSup_eq, hP, sSup_empty] at hs
+
+lemma ne_bot_of_mem (hx : x ∈ P) : x ≠ ⊥ :=
+  fun h ↦ P.bot_not_mem <| h ▸ hx
+
+lemma bot_lt_of_mem (hx : x ∈ P) : ⊥ < x :=
+  bot_lt_iff_ne_bot.2 <| P.ne_bot_of_mem hx
+
+/-- Convert a `Partition s` into a `Partition t` via an equality `s = t`. -/
+@[simps]
+protected def copy {t : α} (P : Partition s) (hst : s = t) : Partition t where
+  parts := P.parts
+  indep := P.indep
+  bot_not_mem := P.bot_not_mem
+  sSup_parts_eq := hst ▸ P.sSup_eq
+
+@[simp]
+lemma coe_copy_eq {t : α} {P : Partition s} (hst : s = t) : (P.copy hst : Set α) = P := rfl
+
+@[simp]
+lemma mem_copy_iff {t x : α} {P : Partition s} (hst : s = t) : x ∈ P.copy hst ↔ x ∈ P := Iff.rfl
+
+/-- The natural equivalence between the subtype of parts and the subtype of parts of a copy. -/
+@[simps!]
+def partscopyEquiv {t : α} (P : Partition s) (hst : s = t) : ↥P ≃ ↥(P.copy hst) :=
+  Equiv.Set.ofEq rfl
+
+/-- A constructor for `Partition s` that removes `⊥` from the set of parts. -/
+@[simps]
+def removeBot (P : Set α) (indep : sSupIndep P) (sSup_eq : sSup P = s) : Partition s where
+  parts := P \ {⊥}
+  indep := indep.mono diff_subset
+  bot_not_mem := by simp
+  sSup_parts_eq := by simp [← sSup_eq]
+
+@[simp]
+lemma coe_removeBot_eq (P : Set α) (indep) (sSup_eq : sSup P = s) :
+    (removeBot P indep sSup_eq : Set α) = P \ {⊥} := rfl
```

Comment:
Hmm, in principle `simps` should be making this lemma itself. What does it give instead?

Reply:
It gives `(removeBot P indep sSup_eq).parts = P \ {⊥} `

Reply:
I think this is a sign that the simps configuration for the new declaration is bad.
I'd hope that the fix of this is something like ilnes 45-48 from https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Topology/Sets/Compacts.lean, when put just after the SetLike instance, will make simps work here.

Reply:
Ok, I think I fixed this.